pkg/types/query/constructors.go

```diff
@@ -0,0 +1,328 @@
+package query
+
+import (
+	"strconv"
+	"time"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/types/query/primitives"
+)
+
+type ComparatorType string
+
+const (
+	AddressComparator  ComparatorType = "ADDRESS"
+	EventSigComparator ComparatorType = "EVENT"
+	TopicComparator    ComparatorType = "TOPIC"
+)
+
+// IndexedLogsKeyFilter creates a KeyFilter that filters logs for the provided topic values at the specified
+// topic index. Topic value filters are 'OR'ed together.
+func IndexedLogsKeyFilter(
+	address, eventSig string,
+	topicIdx uint64,
+	topicValues []string,
+	confidence primitives.ConfidenceLevel,
+) KeyFilter {
+	return KeyFilter{
+		Key: strconv.FormatUint(topicIdx, 10),
+		Expressions: []Expression{
+			addressComparator(address),
+			eventComparator(eventSig),
+			filtersForTopics(topicValues),
+			Confidence(confidence),
+		},
+	}
+}
+
+// IndexedLogsByBlockRangeKeyFilter creates a KeyFilter that filters logs for the provided topic values at the specified
+// topic index. Topic value filters are 'OR'ed together and results are limited by provided block range.
+func IndexedLogsByBlockRangeKeyFilter(
+	addr, eventSig string,
+	start, end uint64,
+	topicIdx uint64,
+	topicValues []string,
+) KeyFilter {
+	return KeyFilter{
+		Key: strconv.FormatUint(topicIdx, 10),
+		Expressions: []Expression{
+			addressComparator(addr),
+			eventComparator(eventSig),
+			filtersForTopics(topicValues),
+			Block(start, primitives.Gte),
+			Block(end, primitives.Lte),
+		},
+	}
+}
+
+// IndexedLogsTopicGreaterThanKeyFilter creates a KeyFilter that filters logs for the provided topic value and index
+// at or above the specified confidence level.
+func IndexedLogsTopicGreaterThanKeyFilter(
+	addr, eventSig string,
```

Comment:
I believe this helpers violate the chain agnostic aspect of `ContractReader`.
Because with this changes cl-common now exposes evm specific details, such as `event sig, logs, log idx, ...`. This means that the caller should be aware of this information, for example CCIP logic should have a reference to an event sig, but this event sig will not exist on two different chain fams.

tl;dr I like the idea of having this helpers but I believe it will lead to chain-specific product implementations and violation of the whole purpose of `ContractReader`.

On the other hand though, If this makes sense and this is something that products will have to do, maybe should at least be under a package that indicates it's `evm` related helpers.

Reply:
I changed from event sig to sequence signature (which identifies the sequence type), topic index to sequence value name (which defines the indexed property name), and sequence value which is the filtered value in the defined property to search. I made some changes to display this in the followup commit. Let me know your thoughts.

Reply:
I do see some issues in the data word index filters as these appear to be doing filtering on raw log data, which would not be consistent between chains.